src/ringct/bulletproofs.cc

```diff
@@ -0,0 +1,750 @@
+{
+  init_exponents();
+
+  PERF_TIMER_UNIT(PROVE, 1000000);
+
+  constexpr size_t logN = 6; // log2(64)
+  constexpr size_t N = 1<<logN;
+
+  rct::key V;
+  rct::keyV aL(N), aR(N);
+
+  PERF_TIMER_START_BP(PROVE_v);
+  rct::addKeys2(V, gamma, sv, rct::H);
+  PERF_TIMER_STOP(PROVE_v);
+     
+  PERF_TIMER_START_BP(PROVE_aLaR);
+  for (size_t i = N; i-- > 0; )
+  {
+    if (sv[i/8] & (((uint64_t)1)<<(i%8)))
+    {
+      aL[i] = rct::identity();
+    }
+    else
+    {
+      aL[i] = rct::zero();
+    }
+    sc_sub(aR[i].bytes, aL[i].bytes, rct::identity().bytes);
+  }
+  PERF_TIMER_STOP(PROVE_aLaR);
+
+     
+  // DEBUG: Test to ensure this recovers the value
+#ifdef DEBUG_BP
+  uint64_t test_aL = 0, test_aR = 0;
+  for (size_t i = 0; i < N; ++i)
+  {
+    if (aL[i] == rct::identity())
+      test_aL += ((uint64_t)1)<<i;
+    if (aR[i] == rct::zero())
+      test_aR += ((uint64_t)1)<<i;
+  }
+  uint64_t v_test = 0;
+  for (int n = 0; n < 8; ++n) v_test |= (((uint64_t)sv[n]) << (8*n));
+  CHECK_AND_ASSERT_THROW_MES(test_aL == v_test, "test_aL failed");
+  CHECK_AND_ASSERT_THROW_MES(test_aR == v_test, "test_aR failed");
+#endif
+     
+  PERF_TIMER_START_BP(PROVE_step1);
+  // PAPER LINES 38-39
+  rct::key alpha = rct::skGen();
+  rct::key ve = vector_exponent(aL, aR);
+  rct::key A;
+  rct::addKeys(A, ve, rct::scalarmultBase(alpha));
+     
+  // PAPER LINES 40-42
+  rct::keyV sL = rct::skvGen(N), sR = rct::skvGen(N);
+  rct::key rho = rct::skGen();
+  ve = vector_exponent(sL, sR);
+  rct::key S;
+  rct::addKeys(S, ve, rct::scalarmultBase(rho));
+     
+  // PAPER LINES 43-45
+  rct::keyV hashed;
+  hashed.push_back(A);
+  hashed.push_back(S);
+  rct::key y = rct::hash_to_scalar(hashed);
+  rct::key z = rct::hash_to_scalar(y);
+     
+  // Polynomial construction before PAPER LINE 46
+  rct::key t0 = rct::zero();
+  rct::key t1 = rct::zero();
+  rct::key t2 = rct::zero();
+           
+  const auto yN = vector_powers(y, N);
+
+  rct::key ip1y = inner_product(oneN, yN);
+  rct::key tmp;
+  sc_muladd(t0.bytes, z.bytes, ip1y.bytes, t0.bytes);
+
+  rct::key zsq;
+  sc_mul(zsq.bytes, z.bytes, z.bytes);
+  sc_muladd(t0.bytes, zsq.bytes, sv.bytes, t0.bytes);
+
+  rct::key k = rct::zero();
+  sc_mulsub(k.bytes, zsq.bytes, ip1y.bytes, k.bytes);
+
+  rct::key zcu;
+  sc_mul(zcu.bytes, zsq.bytes, z.bytes);
+  sc_mulsub(k.bytes, zcu.bytes, ip12.bytes, k.bytes);
+  sc_add(t0.bytes, t0.bytes, k.bytes);
+
+  // DEBUG: Test the value of t0 has the correct form
+#ifdef DEBUG_BP
+  rct::key test_t0 = rct::zero();
+  rct::key iph = inner_product(aL, hadamard(aR, yN));
+  sc_add(test_t0.bytes, test_t0.bytes, iph.bytes);
+  rct::key ips = inner_product(vector_subtract(aL, aR), yN);
+  sc_muladd(test_t0.bytes, z.bytes, ips.bytes, test_t0.bytes);
+  rct::key ipt = inner_product(twoN, aL);
+  sc_muladd(test_t0.bytes, zsq.bytes, ipt.bytes, test_t0.bytes);
+  sc_add(test_t0.bytes, test_t0.bytes, k.bytes);
+  CHECK_AND_ASSERT_THROW_MES(t0 == test_t0, "t0 check failed");
```

Comment:
If this debug code is to be activated under some circumstance, I think this shouldn't throw here, because otherwise the unit test fails for the intended failure cases.

Reply:
I don't think this is an intended failure case, it checks the code above is correct.

Reply:
This is what I get when `DEBUG_BP` is activated:
```
[----------] 4 tests from bulletproofs
[ RUN      ] bulletproofs.valid_zero
[       OK ] bulletproofs.valid_zero (211 ms)
[ RUN      ] bulletproofs.valid_max
[       OK ] bulletproofs.valid_max (236 ms)
[ RUN      ] bulletproofs.invalid_8
2017-12-08 09:22:35.557	  0x7fffae0623c0	ERROR	bulletproofs	src/ringct/bulletproofs.cc:395	t0 check failed
unknown file: Failure
C++ exception with description "t0 check failed" thrown in the test body.
[  FAILED  ] bulletproofs.invalid_8 (54 ms)
[ RUN      ] bulletproofs.invalid_31
2017-12-08 09:22:35.614	  0x7fffae0623c0	ERROR	bulletproofs	src/ringct/bulletproofs.cc:395	t0 check failed
unknown file: Failure
C++ exception with description "t0 check failed" thrown in the test body.
[  FAILED  ] bulletproofs.invalid_31 (52 ms)
[----------] 4 tests from bulletproofs (554 ms total)
```

Reply:
I see, that's because I split the function in two so I could monkey with the keyed amount. Will fix.

Reply:
Actually I'll leave this one as is. Just don't run the unit tests with the debug code in. We want to be able to check the verification fails on such a proof.